### ISSUE TITLE
Fix S1854 FP: `case when` syntax not supported when pattern is a constant

### ISSUE THREAD

**Issue by ValentijnMakkenze** [SonarSource/sonar-dotnet#2311]
### Description
We get a false positive for Dead stores should be removed

### Repro steps
```
  private void TextEditorKeydown(object sender, KeyEventArgs e)
  {
    var tab = new string(' ', 2);  // Remove this useless assignment to local variable 'tab'

    switch (e.KeyCode)
    {
    case Keys.Tab:
    {
      e.SuppressKeyPress = true;
      var ultraTextEditor = (UltraTextEditor)sender;
      var caretPos = ultraTextEditor.Editor.SelectionStart;
      ultraTextEditor.Text = ultraTextEditor.Text.Insert(caretPos, tab);  // We use tab here
      ultraTextEditor.Editor.SelectionStart = caretPos + tab.Length;
      ultraTextEditor.ScrollToCaret();
      break;
    }
  …
```

### Expected behavior
No warning

### Actual behavior
Warning: Remove this useless assignment to local variable 'tab'.

### Known workarounds
Mark item as false positive

### Related information
* SonarC# Version 7.11.0.8083
* Visual Studio Version 2017
* If running through the Scanner for MSBuild, its version 4.5.0.1761

**Comment by andrei-epure-sonarsource** [comment]
hi @ValentijnMakkenze . I cannot repro this FP. I don't have the library for `UltraTextEditor`, so I used the following snippet (with SonarC# 7.14). 

```
    public class WindowsFormsExample
    {
        private string TextEditorKeydown(object sender, KeyEventArgs e)
        {
            var tab = new string(' ', 2);  // Remove this useless assignment to local variable 'tab'

            switch (e.KeyCode)
            {
                case Keys.Tab:
                {
                    e.SuppressKeyPress = true;
                    var text = new string(' ', 1);
                    text = text.Insert(1, tab);  // We use tab here
                    return text;
                }
            }
            return "";
        }
    }
```

Could you please provider a smaller reproducer w/o the need of that `UltraTextEditor` library?

**Comment by ValentijnMakkenze** [comment]
Hi Andrei,

My example was reduced too much, it’s the ’case when’ that came after it that triggers the warning without it the warning disappears.

    private void TextEditorKeydown(object sender, KeyEventArgs e)
    {
      var tab = new string(' ', 2);
      switch (e.KeyCode)
      {
        case Keys.Tab:
        {
          Console.WriteLine(tab);
          break;
        }

        case Keys.A when e.Control:
        {
          break;
        }
      }
    }

Kind regards,

Valentijn

Van: Andrei Epure <notifications@github.com>
Verzonden: Tuesday, June 18, 2019 12:13 PM
Aan: SonarSource/sonar-dotnet <sonar-dotnet@noreply.github.com>
CC: Valentijn Makkenze <V.Makkenze@vicrea.nl>; Mention <mention@noreply.github.com>
Onderwerp: Re: [SonarSource/sonar-dotnet] False positive for S1854: Dead stores should be removed (#2311)


hi @ValentijnMakkenze<https://github.com/ValentijnMakkenze> . I cannot repro this FP. I don't have the library for UltraTextEditor, so I used the following snippet (with SonarC# 7.14).

    public class WindowsFormsExample

    {

        private string TextEditorKeydown(object sender, KeyEventArgs e)

        {

            var tab = new string(' ', 2);  // Remove this useless assignment to local variable 'tab'



            switch (e.KeyCode)

            {

                case Keys.Tab:

                {

                    e.SuppressKeyPress = true;

                    var text = new string(' ', 1);

                    text = text.Insert(1, tab);  // We use tab here

                    return text;

                }

            }

            return "";

        }

    }

Could you please provider a smaller reproducer w/o the need of that UltraTextEditor library?

**Comment by andrei-epure-sonarsource** [comment]
Ok, thanks for the follow-up @ValentijnMakkenze. This issue is related to #2010, in the sense that both rules don't support `case when` statements

**Comment by andrei-epure-sonarsource** [comment]
Fixed in #2493
